File: day03/day03.go
package day03

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// Solution for star 1 first attempt
func SolveStar1Naive() int {
	// Read the input file
	file, err := os.Open("day03/input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return -1
	}
	defer file.Close()

	re := regexp.MustCompile(`mul\(\d{1,3},\d{1,3}\)`)
	var toMultiply []string
	// Process the input file line by line
	var scanner = bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		regExMatches := re.FindAllString(line, -1)
		toMultiply = append(toMultiply, regExMatches...)

	}
	//fmt.Println(toMultiply)

	sum := 0
	for _, element := range toMultiply {
		nums := strings.Split(element[4:len(element)-1], ",")
		num1, _ := strconv.Atoi(nums[0])
		num2, _ := strconv.Atoi(nums[1])
		sum += num1 * num2
	}

	return sum
}

// Solution for star 2 first attempt
func SolveStar2Naive() int {
	// Read the input file
	file, err := os.Open("day03/simpleinput.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return -1
	}
	defer file.Close()

	// Process the input file line by line
	var scanner = bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		fmt.Println(line)
	}

	return 0
}

File: day03/day03.py
import re

def SolveStar1():
    pattern = re.compile("(?<=mul\\()\\d{1,3},\\d{1,3}(?=\\))")

    toMultiply = []
    # Open the file in read mode
    with open('day03/input.txt', 'r') as file:
        # Read each line in the file
        for line in file:
            # Print each line
            toMultiply += re.findall(pattern, line)

    total = 0
    for element in toMultiply:
        nums = element.split(',')
        num1 = int(nums[0])
        num2 = int(nums[1])
        total += num1 * num2

    return total


def SolveStar2():
    pattern = re.compile("((?<=mul\\()\\d{1,3},\\d{1,3}(?=\\)))|(don't\\(\\))|(do\\(\\))")

    toMultiply = []
    # Open the file in read mode
    with open('day03/input.txt', 'r') as file:
        # Read each line in the file
        for line in file:
            # Print each line
            toMultiply += re.findall(pattern, line)

    mulEnabled = True
    total = 0
    for instruction in toMultiply:
        # dont
        if instruction[1] != '':
            mulEnabled = False
        elif instruction[2] != '':
            mulEnabled = True
        elif instruction[0] != '' and mulEnabled:
            nums = instruction[0].split(',')
            num1 = int(nums[0])
            num2 = int(nums[1])
            total += num1 * num2
        
    return total